lti_consumer/lti_1p3/consumer.py

```diff
@@ -704,3 +708,178 @@ def enable_nrps(self, context_memberships_url):
+    resource_link, etc. This information is provided to the consumer through the set_proctoring_data method, which
+    is called from the consuming context to pass in necessary data.
+    """
+    def __init__(
+        self,
+        iss,
+        lti_oidc_url,
+        lti_launch_url,
+        client_id,
+        deployment_id,
+        rsa_key,
+        rsa_key_id,
+        tool_key=None,
+        tool_keyset_url=None,
+    ):
+        """
+        Initialize the LtiProctoringConsumer by delegating to LtiConsumer1p3's __init__ method.
+        """
+        super().__init__(
+            iss,
+            lti_oidc_url,
+            lti_launch_url,
+            client_id,
+            deployment_id,
+            rsa_key,
+            rsa_key_id,
+            tool_key,
+            tool_keyset_url
+        )
+        self.proctoring_data = {}
+
+    def set_proctoring_data(self, **kwargs):
+        """
+        Sets the self.proctoring_data dictionary with the provided kwargs, so long as a given key is in
+        LTI_PROCTORING_DATA_KEYS.
+        """
+        for key, value in kwargs.items():
+            if key in LTI_PROCTORING_DATA_KEYS:
+                self.proctoring_data[key] = value
+
+    def _get_base_claims(self):
+        """
+        Returns claims common to all LTI Proctoring Services LTI launch messages, to be used when creating LTI launch
+        messages.
+        """
+        proctoring_claims = {
+            "https://purl.imsglobal.org/spec/lti-ap/claim/attempt_number": self.proctoring_data.get("attempt_number"),
+            "https://purl.imsglobal.org/spec/lti-ap/claim/session_data": self.proctoring_data.get("session_data"),
+        }
+
+        return proctoring_claims
+
+    def get_start_proctoring_claims(self):
+        """
+        Returns claims specific to LTI Proctoring Services LtiStartProctoring LTI launch message,
+        to be injected into the LTI launch message.
+        """
+        proctoring_claims = self._get_base_claims()
+        proctoring_claims.update({
+            "https://purl.imsglobal.org/spec/lti/claim/message_type": "LtiStartProctoring",
+            "https://purl.imsglobal.org/spec/lti-ap/claim/start_assessment_url":
+                self.proctoring_data.get("start_assessment_url"),
+        })
+
+        return proctoring_claims
+
+    def get_end_assessment_claims(self):
+        """
+        Returns claims specific to LTI Proctoring Services LtiEndAssessment LTI launch message,
+        to be injected into the LTI launch message.
+        """
+        proctoring_claims = self._get_base_claims()
+        proctoring_claims.update({
+            "https://purl.imsglobal.org/spec/lti/claim/message_type": "LtiEndAssessment",
+        })
+
+        return proctoring_claims
+
+    def generate_launch_request(
+        self,
+        preflight_response,
+    ):
+        """
+        Builds and return LTI launch message for proctoring.
```

Comment:
Question: Pretty basic, but could you clarify what the LTI launch message is for?